### PR TITLE
Improve downstream lag behavior

### PR DESCRIPTION
**Pull request by Gerold103**
`box.info.replication[replica_id].downstream.lag` shows how much time it took the latest transaction, acked by the given replica, to be replicated to the replica, be written to its WAL, and be acked by this instance.

It sounds simple, but sometimes the actual behaviour was misleading or hard to use.

The patchset improves the user experience. The main change is that the lag is actually updated on each acked txn. Before that it was only updated for acked txns created by this specific master.

See #9748.

### CODE REVIEW

File: test/replication-luatest/gh_9748_downstream_lag_test.lua [hunk 1]
```diff
@@ -63,3 +69,109 @@ g.test_lag_on_master_restart = function(cg)
...
+    -- Server1 makes a txn, replicated to server3 via server2. But that still
+    -- bumps the downstream lag in server2->server3. Downstream lag is updated
+    -- even when the replica confirms third-node txns, not only master's own
+    -- txns.
+    cg.server1:exec(function()
+        box.space.test:replace{2}
+    end)
+    cg.server2:wait_for_downstream_to(cg.server3)
+    cg.server2:exec(function(id, delay)
+        t.assert_le(box.info.replication[id].downstream.lag, delay)
```
Comment (sergepetrenko) [inline test/replication-luatest/gh_9748_downstream_lag_test.lua:100]: Let's also test that the lag is greater than zero

File: test/replication-luatest/gh_9748_downstream_lag_test.lua [hunk 2]
```diff
@@ -63,3 +69,109 @@ g.test_lag_on_master_restart = function(cg)
...
+    cg.server2:exec(function(delay, replication)
+        require('fiber').sleep(delay)
+        box.cfg{replication = replication}
+    end, {delay, {cg.replication[1], cg.replication[2]}})
+    -- server1 -> server2 -> server3
+    cg.server3:update_box_cfg{replication = {cg.replication[2]}}
+    cg.server2:wait_for_downstream_to(cg.server3)
+    -- Server2 has a low lag to server3.
+    cg.server2:exec(function(id, delay)
+        t.assert_le(box.info.replication[id].downstream.lag, delay)
```
Comment (sergepetrenko) [inline test/replication-luatest/gh_9748_downstream_lag_test.lua:133]: Same about lag being greater than zero

**Review by sergepetrenko — APPROVED** [review]
Thanks for the patch!

LGTM. Just a couple nitpicks regarding the test.

File: test/replication-luatest/gh_9748_downstream_lag_test.lua (hunk 1 above)
Comment (Gerold103) [reply test/replication-luatest/gh_9748_downstream_lag_test.lua]: Thanks, added.

File: test/replication-luatest/gh_9748_downstream_lag_test.lua (hunk 2 above)
Comment (Gerold103) [reply test/replication-luatest/gh_9748_downstream_lag_test.lua]: Done.

**Review by Serpentian — APPROVED** [review]
Thank you for the patch. Looks great! One minor comment

File: src/box/applier.cc [hunk 3]
```diff
@@ -1254,8 +1254,8 @@ static void
 replica_txn_wal_write_cb(struct replica_cb_data *rcb)
 {
 	struct replica *r = replica_by_id(rcb->replica_id);
-	if (likely(r != NULL))
-		r->applier_txn_last_tm = rcb->txn_last_tm;
+	if (likely(r != NULL && r->applier != NULL))
+		r->applier->txn_last_tm = rcb->txn_last_tm;
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply src/box/applier.cc:1259]: The problem is that while the txn is in progress, the applier has no guarantee to survive until the txn is complete. AFAIU, this is why we don't just pass applier pointer with each applier-originated transaction.

I didn't check that this is the case specifically, but from the code it looks possible (for applier to be deleted while itx txn is in progress).

I wouldn't mind spending a bit more time on that though. The applier surely could be simplified further, at least in this regard. Should I do that in scope of this PR, or work on it separately, @sergepetrenko?

**Review by xuniq — APPROVED** [review]

File: changelogs/unreleased/gh-9748-downstream-lag-bug.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+## bugfix/replication
+
+* Fixed a bug that `box.info.replication[...].downstream.lag` value could be
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply changelogs/unreleased/gh-9748-downstream-lag-bug.md]: Thanks, fixed.

File: src/box/applier.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/applier.cc:1259]: I suggest doing this in a separate PR.
Let's just fix the bug here without too much refactoring.

[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/applier.cc:1259]: @Serpentian please reassign me if you're fine with current patch version.

**Comment by sergepetrenko** [comment]
Backports:
3.1 https://github.com/tarantool/tarantool/pull/10024
2.11 https://github.com/tarantool/tarantool/pull/10025
